### PR TITLE
[VideoVersion] Improved the same video detection

### PR DESCRIPTION
**Pull request by xodidox**
## Description

Improved the same video detection.

Currently if both videos have same title, they will be detected as same video, this change will only detect two videos as same video if:
1. they have same title and year
2.  they have same title but both have no year

Just for completeness, which is already in the current implementation. If both videos have same unique ids (imdb id/tmdb id, etc.), they will be detected as same video, not matter other conditions.
## Motivation and context

Feedback
## How has this been tested?

Runtime tested on MacBook
## What is the effect on users?

Reduced the false positive for same video detection.
## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by 78andyp** [comment]
Brilliant, thanks. Works as expected. Will close mine.

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [inline xbmc/video/VideoDatabase.cpp:3996]: wouldn't just the condition `year == currentYear` basically cover all the other conditions?

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [inline xbmc/video/VideoDatabase.cpp:3990]: I am a little concerned about all the additional DB queries that are done in the background here and the performance impact they might have on the scraping process, especially when you do an initial import of your library.

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3996]: There are APIs for this, is it a good practice to use the implement details here?

File: xbmc/video/VideoDatabase.cpp
Comment (the-black-eagle) [reply xbmc/video/VideoDatabase.cpp:3990]: Don't need a join.  You could just do `SELECT idMovie, substring(Premiered,1,4) as year FROM movie WHERE movie..c%02d = '%s'", VIDEODB_ID_TITLE, title.c_str()));`

AFAIK, it is preferable to fetch as much of the needed info from the db with as few calls as possible.

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:3996]: both methods have their use depending on the context they are needed. But here you only care if `year == year` or if `no year == no year` and those two cases are already covered by `year == year` because `getYear()` internally does the same checks as `hasYear()` and returns 0 if no year is present. So the complex condition could IMO be simplified, and if it is only to safe a few CPU cycles per video. That would at least be my take on it - not sure how others see this (I am by no means any code authority)

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:3990]: > Good points, yes there is an impact for external databases, although in this case there probably wouldn't be a lot of matches for one movie name and therefore not many extra requests. But still...

if you scan a library with thousands of movies, queries add up. And our scraping aready isn't the fastest anyways

Comment (the-black-eagle) [reply xbmc/video/VideoDatabase.cpp:3990]: The movie table only has a premiered date. 

https://github.com/xbmc/xbmc/blob/master/xbmc/video/VideoDatabase.cpp#L4285

You can do it all in the sql query.  `select idmovie ,case when instr('-',movie.premiered) > 5 then substring(premiered,7,4) else substring(premiered,1,4) end as year from movie where movie.c%02d = '%s'", VIDEODB_ID_TITLE, title.c_str()));`  

I think that will handle yyyy-mm-dd, mm-dd-yyyy and just yyyy for both MySQL/mariadb & sqlite

Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:3990]: just a thought, but couldn't we already check for a matching year in the query? So either one complex query covering all cases (incoming movies has year or not), or a dedicated query for `incoming movie has year` and `incoming movies has no year`. That way we eliminate all the other overhead

Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3990]: For library scanning, the time-consuming part should be file system overhead and downloading from Internet, is it really necessary to go with above ways?

Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:3990]: performance overhead certainly matters, especially for low powered devices like RPI and other SBCs. And if you have a look at `GetDetailsByTypeAndId` it will at least trigger two or three additional queries (especially since this method will also trigger `HasVideoVersions()`, for which we already have a complaint that this feature noticeably slows down the listing of larger movie libraries which we also have to address)

Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3990]: I'm not saying the perf overhead is not matter, I'm saying in comparison with the big factors of perf, is this really necessary? The previous complaint about perf was because the old code tried to query DB for a second time during media window update, now I cached this information in VideoInfoTag, and didn't see any perf complaint currently. Anyway, I can cache the GetDetailsByTypeAndId result as you explained in OP as a solution.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3996]: This comparison (and orig PR code) compares the year and ignores the month/day part when it's available (premiered date with format YYYY-MM-DD instead of just year YYYY)

Should the compared items be required to both have premiered date or to both have only year? Or is it too strict?
I don't know much about scrapers, nfo or the db and it could be a stupid question ;-)
A comment would be nice if the month/day part is ignored on purpose.

For example is it the same item if the DB item has the premiered date 2023-11-28 and is compared with item that has only the year 2023?

**Comment by xodidox** [comment]
@da-anda Updated.

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [inline xbmc/video/VideoDatabase.cpp:3945]: sorry, haven't noticed before, but this query seems fishy or I just don't understand what you are trying to do here. Why do you need a subselect on the table you query in your main query? Wouldn't this query return the same result?
```
"SELECT DISTINCT media_id "
"FROM uniqueid "
"WHERE media_id = %i AND media_type = '%s')"
```

Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3945]: There is comment for this above the query. It's used to get all media ids which have the same unique ids as the specified media id. What your query above just returns the media id you specified. You can check the uniqueid table to see how it works.

Comment (the-black-eagle) [reply xbmc/video/VideoDatabase.cpp:3945]: You can't have an item with the same unique id, else it wouldn't be unique!  What the query fetches is actually a distinct value for the media id.  However as you are already specifying a media_id it can only ever fetch that specified value.  EG, if I have three movie versions that have a media_id (in the uniqueid table) of say 55, then a query of 
`SELECT distinct media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 55 AND media_type = 'movie')`  will just return one row of 55, irrespective of the fact that there are 3 versions.  The uniqueid bit doesn't seem to come in to it.

@da-anda , do you agree ?

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:3995]: Probably more readable if integrated in the cursor loop just above, to add to the vector only when id isn't known yet. Would remove one level of iterator and save a few cycles.

Comment (the-black-eagle) [reply xbmc/video/VideoDatabase.cpp:3995]: I'll try and explain better then!

The uniqueid query uses a media_id to select some values from the uniqueid table.  It then uses those values to select a media_id from the same table.  Because it all operates on the same table, the resultant media_id MUST always be the same as the one we supplied in the first place.

The query operates by saying "give me all the imdb & tmdb id's for the movie with id 'X'"  It then says, "Which movie has these imdb and tmdb id's?"  The answer is obviously movie 'X' every time.

Actual output from a live db

|        7900 |     1865 | movie      | 504172     | tmdb                    |
|        7901 |     1866 | movie      | tt1211837  | imdb                    |
|        7902 |     1866 | movie      | 284052     | tmdb                    |
|        7903 |     1867 | movie      | tt1764651  | imdb                    |
|        7904 |     1867 | movie      | 76163      | tmdb                    |
|        7905 |     1868 | movie      | tt0332379  | imdb                    |
|        7906 |     1868 | movie      | 1584       | tmdb                    |
|        7907 |     1869 | movie      | tt0079574  | imdb                    |
|        7908 |     1869 | movie      | 698        | tmdb                    |
|        7909 |     1870 | movie      | tt0097742  | imdb                    |
|        7910 |     1870 | movie      | 709        | tmdb                    |
|        7911 |     1871 | movie      | tt1727824  | imdb                    |
|        7912 |     1871 | movie      | 424694     | tmdb                    |
|        7913 |     1872 | movie      | tt0120737  | imdb                    |
|        7914 |     1872 | movie      | 120        | tmdb                    |
|        7915 |     1873 | movie      | tt0082398  | imdb                    |
|        7916 |     1873 | movie      | 699        | tmdb                    |
|        7917 |     1874 | movie      | tt2247732  | imdb                    |
|        7918 |     1874 | movie      | 119569     | tmdb                    |
+-------------+----------+------------+------------+-------------------------+
2370 rows in set (0.001 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1867 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1867 |
+----------+
1 row in set (0.005 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1871 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1871 |
+----------+
1 row in set (0.005 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1873 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1873 |
+----------+
1 row in set (0.005 sec)

File: xbmc/video/VideoDatabase.cpp
Comment (da-anda) [reply xbmc/video/VideoDatabase.cpp:3977]: I don't think that we have a limit on indices. So adding a new index for the title should be fine

**Comment by the-black-eagle** [comment]
@xodidox Ah, I see what you are trying to do here, and also why I wasn't getting the results you are obviously seeing.  My library is built using local nfo files, but many of these have not been updated since 2014 and although they do contain imdb and/or tmdb id's these are not in the tags that current nfo files use.  So, they get added to the uniqueid table as a type of 'unknown'.  Obviously then, the type doesn't match a later added movie with the same id, even though those two imdb id's do match.

So, somehow you need to account for db's that have been upgraded and not re-scraped, or have been built from older data as in both cases 'type' may be unknown.

The 1st movie in my db is Avatar.  The corresponding part of the nfo file is `<id>tt0499549</id>
    <tmdbid>19995</tmdbid>`

This is stored in the uniqueid table as 
1 | 1 | movie | tt0499549 | unknown
-- | -- | -- | -- | --

I also have a 4k version of it, but the nfo file was created this year and does contain the correct tags.  However, for the reason explained above, the current query fails to find two entries in the db.

140 | 120 | movie | tt0499549 | imdb
-- | -- | -- | -- | --
141 | 120 | movie | 19995 | tmdb

**Comment by xodidox** [comment]
If unique ids don't match and movie title don't match either, you are on your own. You can convert one of them as the additional version of another from the VideoVersion manage dialog, use "Add version" button to add the video file directly, the movie entry will be deleted automatically with a confirmation. I don't think use a hack (checking if it's IMDb id) is the way to go.

**Comment by xodidox** [comment]
@da-anda where can I find the data format of the movie.premiered field, thus can confirm the first 4 chars are year? From the code that refer this seems not in this fixed format. I'll take this kind of hack as last sort if there is big perf impact for the change, as there is API HasYear and GetYear for this.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3945]: The comment is technically correct but doesn't help much. Maybe something like "Retrieve items that share a metadata identifier: alternative versions imported before versions feature, or a version merge was rejected by the user"

Yes @da-anda the `WHERE (value, type) IN (SELECT value, type...` eliminates the rows that don't match both fields.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3995]: my original comment was missed I think, the uniqueid table was not the point.

**Comment by xodidox** [comment]
> I'd say yes, in the context of this function it's probably better to show an extra match rather than omit one?

The intention is to make the match as accurate as possible. A wrong match will bug user with an unexpected pop dialog to confirm version selection.

**Comment by xodidox** [comment]
@da-anda @CrystalP Updated with the suggested query.

**Comment by CrystalP** [comment]
> The intention is to make the match as accurate as possible. A wrong match will bug user with an unexpected pop dialog to confirm version selection.

OK I misunderstood the flow. I thought that this was also used to present those most likely good matches at the top of the list of movies to pick from when adding a version for an already imported movie. But it's not, the list is all the movies of the library.

**Comment by da-anda** [comment]
my only remark now would be that with the adjusted query the code could be simplified by a lot by merging loops etc. So instead of

1. aggregate IDs with matching uniqueid
2. loop over IDs, fetch details and add them to the `items` list
3. aggregate IDs with matching title+year
4. loop over IDs, fetch details and add them to the ` items`  list

it basically could also be

1. aggregate IDs with matching uniqueid
2. aggregate IDs with matching title+year and add them to the ID list
3. loop over IDs, fetch details and add them to the ` items`  list

so basically like this maybe?
```
void CVideoDatabase::GetSameVideoItems(CFileItem& item, CFileItemList& items)
{
  if (!m_pDB || !m_pDS)
    return;

  std::vector<int> itemIds;

  int dbId = item.GetVideoInfoTag()->m_iDbId;
  MediaType mediaType = item.GetVideoInfoTag()->m_type;

  try
  {
    // get items with same unique ids (imdb, tmdb, etc)
    m_pDS->query(
        PrepareSQL("SELECT DISTINCT media_id "
                   "FROM uniqueid "
                   "WHERE (value, type) IN "
                   "  (SELECT value, type FROM uniqueid WHERE media_id = %i AND media_type = '%s')",
                   dbId, mediaType.c_str()));

    while (!m_pDS->eof())
    {
      itemIds.push_back(m_pDS->fv("media_id").get_asInt());
      m_pDS->next();
    }

    m_pDS->close();

    VideoDbContentType itemType = item.GetVideoContentType();

    // get items matching same title and year
    if (itemType == VideoDbContentType::MOVIES)
    {
      if (item.GetVideoInfoTag()->HasYear())
        m_pDS->query(PrepareSQL("SELECT idMovie FROM movie WHERE movie.c%02d = '%s' AND "
                                "SUBSTR(movie.premiered, 1, 4) = '%s'",
                                VIDEODB_ID_TITLE, item.GetVideoInfoTag()->GetTitle().c_str(),
                                item.GetVideoInfoTag()->GetYear()));
      else
        m_pDS->query(
            PrepareSQL("SELECT idMovie FROM movie WHERE movie.c%02d = '%s' AND NOT movie.premiered",
                       VIDEODB_ID_TITLE, item.GetVideoInfoTag()->GetTitle().c_str()));

      while (!m_pDS->eof())
      {
        int movieId = m_pDS->fv("idMovie").get_asInt();
        // only add movieId if not already in itemIds
        if (std::find(itemIds.begin(), itemIds.end(), movieId) == itemIds.end())
          itemIds.push_back(movieId);
        m_pDS->next();
      }

      m_pDS->close();
    }

    // add items to list
    for (const auto id : itemIds)
    {
      auto current = std::make_shared<CFileItem>();
      if (GetDetailsByTypeAndId(*current.get(), itemType, id))
        items.Add(current);
    }
  }
  catch (...)
  {
    CLog::Log(LOGERROR, "{} failed for {} {}", __FUNCTION__, mediaType, dbId);
  }
}
``` 
the code is untested

**Comment by xodidox** [comment]
@da-anda Updated.

**Review by da-anda — APPROVED** [review]
thanks, looks good to me now. @CrystalP your call to push the button if you think the code is ok now

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [inline xbmc/video/VideoDatabase.cpp:3968]: I'm not convinced yet that `movie.premiered LIKE '%i%%'` is best but let's go with that for now, it's used in many places already.
movie.premiered is text not a bool, I think testing movie.premiered = '0' is more correct.

**Comment by CrystalP** [comment]
Not sure how important and whether to hold: this doesn't seem to deal with unicode characters in the title, but I'm new to the db and don't know how it's dealt with for other fields. Most parts of Kodi I know use UTF16, not UTF8 so a c_str() raises a flag.

Taking a step back, and I'm learning about the existing code as we go, for future changes we should probably rely more on existing advanced functions like GetMoviesByWhere() and extend the CFilter as needed, to grab all the information as one batch instead of having to repeat the same cursor loops all over the place.

**Comment by the-black-eagle** [comment]
> Not sure how important: this doesn't seem to deal with unicode characters in the title, but I'm new to the db and don't know how it's dealt with for other fields. Most parts of Kodi I know use UTF16, not UTF8 so a c_str() raises a flag.

c_str() is correct.  However, when I wrote the videolinks stuff, I originally wrote it to match on text (song name).  This proved to be unworkable due to differences between Linux and Windows string handling.  In spite of a particular song being tagged by the same tagger (using the same online source), the Win version refused to find the match where the Linux version did.  This is documented in the associated forum thread.  

In summary, although matching with text _should_ work, in practice, there appears to be cross-platform differences that can interfere with that.  I'm not saying that this is actually the case here, it's just something to be aware of going forwards. 

Matching on imdb/tmdb id's should always be the preferred method, although I still think the code should allow for `unknown` types as well.

> Taking a step back, and I'm learning about the existing code as we go, for future changes we should probably rely more on existing advanced functions like GetMoviesByWhere() and extend the CFilter as needed, to grab all the information as one batch instead of repeating the same cursor loops all over the place.

Fully agree.  `GetMoviesByWhere()` is indeed the way it should work, in a similar manner as to how set members are retrieved.

**Comment by xodidox** [comment]
Please check the PR description to get a whole picture of the change. We're always using unique ids (imdb/tmdb, etc.) to get matches, but for cases that no unique ids, we're making a work around here to match the potential same movies with title (and/or year if exists). If the work around doesn't work, user can always find the movie by themselves and convert it to an additional version from context menu. The miss for the latter case is good, a wrong match is bad here as it will pop up an unexpected dialog to ask user to confirm.

**Comment by CrystalP** [comment]
I guess we'll take a chance with unicode for now. @the-black-eagle I would appreciate a link to the forum thread you mentioned, there should be a way to solve.

@xodidox missing for me in current revision: index creation, text comparison with year 0.

'unknown' metadata source: I tend to agree with @da-anda's earlier comments that the identifiable ones should be migrated, and the rest is ambiguous and not to be used. It's still possible to manually make a newly imported file into a new version of an existing.

Better handling unknown metadata source would be another project (maybe a scheduled script?) Now that I've looked, I also have a bunch in my db, that I will fix by script.

File: xbmc/video/VideoDatabase.cpp
Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3968]: @CrystalP I tested the premiered column, seems an empty string works too, is "0" the default value for this column? what about using "LENGTH(premiered) < 4" as a correct premiered column length will be at least 4?

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3968]: premiered is text. no default that I can see in code.

The field is serialized by `CVideoDatabase::SetDetailsForMovie()` / `UpdateDetailsForMovie()` with
```C++
    if (details.HasPremiered())
      sql += PrepareSQL(", premiered = '%s'", details.GetPremiered().GetAsDBDate().c_str());
    else
      sql += PrepareSQL(", premiered = '%i'", details.GetYear());
```
Follow the code, looks to me like any null, blank or negative values of premiered would be bugs.
0 is the value that means "unknown". I only found rows with 0, YYYY or YYYY-MM-DD in my db.
In code, `GetYear()` returns 0 when invalid instead of the -1 it may contain internally. Same with premiered, returns 0 when invalid. All the pieces fit for me.

Comparison with empty string doesn't work here with sqlite, as expected because '' is not the same string as '0'

Comment (xodidox) [reply xbmc/video/VideoDatabase.cpp:3968]: This is how the column is interpreted in code:

CVideoDatabase::GetDetailsForMovie:

  if (premieredString.size() == 4)
    details.SetYear(record->at(VIDEODB_DETAILS_MOVIE_PREMIERED).get_asInt());
  else
    details.SetPremieredFromDBDate(premieredString);

where the SetPremieredFromDBDate method will use CDateTime::SetFromDBDate to interpret the value

bool CDateTime::SetFromDBDate(const std::string &date)
{
  if (date.size() < 10)
    return false;

So the length of this column is at least 4, less than that will be invalid. I can't find anywhere refer the "0" string. If there is no concern from you, I'll go with the "LENGTH(premiered) < 4" way.

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3968]: Fine. It's getting silly and for this simple case many tests work.

You don't find the '0' in code because, it's the 0 integer value returned by `GetYear()` for invalid dates, converted to text.

File: xbmc/video/VideoDatabase.cpp
Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:3945]: Please provide a more descriptive comment for our future selves, see couple comments above.

**Comment by xodidox** [comment]
@CrystalP Updated based on feedback.

**Review by CrystalP — APPROVED** [review]
Good change to avoid matching a movie on title only when metadata id is not available.

File: xbmc/video/VideoDatabase.cpp
Comment (ksooo) [inline xbmc/video/VideoDatabase.cpp:293]: Just asking, will the index created without a database version bump?

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:293]: no you're right it won't...

Comment (CrystalP) [reply xbmc/video/VideoDatabase.cpp:293]: blessing in disguise actually, the index is fine in sqlite but based on previous PR mariadb probably won't like it without field length.

**Pull request merged**
